### PR TITLE
feat: scaling factors validation in create stableswap pool, small refactor and unit tests

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #1392

## What is the purpose of the change

Implemented scaling factors for `CreateStableswapPool`, `NewStableswapPool`, along with `ValidateBasic` validation.

Refactored to make it more testable and align the design closer to balancer.

All changed logic is covered by unit tests.

## Brief Changelog

- [introduce PoolAssets in the stableswap pool to make more consistent with the balancer](https://github.com/osmosis-labs/osmosis/pull/1459/commits/2e6aa311798943743509f07dc85e0aba67969d0a) 
- [implement TestStableswapCreatePool](https://github.com/osmosis-labs/osmosis/pull/1459/commits/de5eac4503c8096a88bad805c710dc02a06ddb79)
- [add TestStableswapGetTotalPoolLiquidity unit test](https://github.com/osmosis-labs/osmosis/pull/1459/commits/890cb71cc484eb3859812c5543b2e48345e1b68f)
- [stableswap pool validation; getScaledPoolAmt; unit tests](https://github.com/osmosis-labs/osmosis/pull/1459/commits/57ac345722bf07001ac0d69f10a03e4b8a4a6ed8)
- [TestStableswapGetDeScaledPoolAmt unit test](https://github.com/osmosis-labs/osmosis/pull/1459/commits/028a6728ffbdb4a39eee353ed43524dd17074142)
- [implement TestStableswapGetPoolAssetAndIndex](https://github.com/osmosis-labs/osmosis/pull/1459/commits/307cb7260204474cb1898ae4470fb1f3fb96142f)
- [implement TestStableswapUpdatePoolLiquidityForSwap](https://github.com/osmosis-labs/osmosis/pull/1459/commits/a447331cce3d3134023a6dcfc9b2f9a81d7e3c9c)
- [pool asset validation in message; TestStableSwapPoolAssetValidate](https://github.com/osmosis-labs/osmosis/pull/1459/commits/4046428d0419872edc23d349dbb4ff16eb1fbcac)
- [extract constants for denom names in tests](https://github.com/osmosis-labs/osmosis/pull/1459/commits/ec65196cc53d455b029a3dc19439a44133a496b0)

## Testing and Verifying

This change introduces a scaling factors to CreateSta

includes a minor refactor to introduce `PoolAssets` in stableswap pool to make the design more aligned with the balancer and more testable. 

This change added tests and can be verified as follows:
```
go test github.com/osmosis-labs/osmosis/v7/x/gamm/pool-models/stableswap
```

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? yes - proto files were changed
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not applicable

### CODE REVIEW

File: x/gamm/pool-models/stableswap/msgs.go [hunk 1]
```diff
@@ -86,14 +76,44 @@ func (msg MsgCreateStableswapPool) Validate(ctx sdk.Context) error {
...
 		return nil, err
 	}
 
-	return &stableswapPool, nil
+	return stableswapPool, nil
+}
+
+func ValidatePoolAssets(poolAssets []PoolAsset) error {
+	// validation for pool initial liquidity
+	// TODO: expand this check to accommodate multi-asset pools for stableswap
```
Comment (p0mvn) [inline x/gamm/pool-models/stableswap/msgs.go:101]: this was copied over, what are the plans for this?

File: x/gamm/pool-models/stableswap/pool.go [hunk 2]
```diff
@@ -154,7 +139,9 @@ func (pa *Pool) SwapOutAmtGivenIn(ctx sdk.Context, tokenIn sdk.Coins, tokenOutDe
 		return sdk.Coin{}, err
 	}
 
-	pa.updatePoolLiquidityForSwap(tokenIn, sdk.NewCoins(tokenOut))
+	if err := pa.updatePoolLiquidityForSwap(tokenIn, sdk.NewCoins(tokenOut)); err != nil {
+		panic(err)
```
Comment (p0mvn) [inline x/gamm/pool-models/stableswap/pool.go:143]: this was changed to make it more testable in unit tests

File: x/gamm/pool-models/stableswap/pool.go [hunk 3]
```diff
@@ -219,3 +215,50 @@ func (pa Pool) CalcExitPoolShares(ctx sdk.Context, numShares sdk.Int, exitFee sd
 
 // no-op for stableswap
 func (pa *Pool) PokePool(blockTime time.Time) {}
+
+// Returns a pool asset, and its index. If err != nil, then the index will be valid.
+// CONTRACT: pool must be created with NewStableSwapPool so that the pool assets are sorted
+// by denom. Otherwise, the behavior is undefined.
+func (pa Pool) getPoolAssetAndIndex(denom string) (int, PoolAsset, error) {
```
Comment (p0mvn) [inline x/gamm/pool-models/stableswap/pool.go:222]: this can be later generalized wit balancer

**Comment by ValarDragon** [comment]
I think the non-pool asset design was better tbh, and Balancer is what needs changing in the future

**Comment by p0mvn** [comment]
> I think the non-pool asset design was better tbh, and Balancer is what needs changing in the future

Why do you think so?

I find it more brittle to rely on the order of 2 separate slices to be in sync. If 2 slices need to maintain the same order, they should be grouped and merged into one. Otherwise, a small update to the ordering of one of the slices in the future can bring them out of sync and might cause problems when debugging.

Additionally, while setting up unit tests, I found this abstraction to be easier to configure and assert on its results.

These were the 2 primary motivators for this change

Please let me know what you think

File: proto/osmosis/gamm/pool-models/stableswap/stableswap_pool.proto [hunk 4]
```diff
@@ -29,6 +29,22 @@ message PoolParams {
...
+// Pool asset is an internal struct that combines the amount of the
+// token in the pool, and scaling factor.
+message PoolAsset {
+  // Coins we are talking about,
+  // the denomination must be unique amongst all PoolAssets for this pool.
+  cosmos.base.v1beta1.Coin token = 1
+      [ (gogoproto.moretags) = "yaml:\"token\"", (gogoproto.nullable) = false ];
+  // used for normalizing values for calculation amognst assets with different
+  // precisions
+  string scalingFactor = 2 [
```
Comment (alexanderbez) [inline proto/osmosis/gamm/pool-models/stableswap/stableswap_pool.proto:41]:
```suggestion
  string scaling_factor = 2 [
```

File: proto/osmosis/gamm/pool-models/stableswap/stableswap_pool.proto [hunk 5]
```diff
@@ -59,13 +75,6 @@ message Pool {
...
-  repeated cosmos.base.v1beta1.Coin poolLiquidity = 6 [
-    (gogoproto.nullable) = false,
-    (gogoproto.castrepeated) = "github.com/cosmos/cosmos-sdk/types.Coins"
-  ];
-  // for calculation amognst assets with different precisions
-  repeated uint64 scaling_factor = 7 [
-    (gogoproto.moretags) = "yaml:\"stableswap_scaling_factor\"",
-    (gogoproto.nullable) = false
-  ];
+  repeated osmosis.gamm.stableswap.v1beta1.PoolAsset poolAssets = 6
```
Comment (alexanderbez) [inline proto/osmosis/gamm/pool-models/stableswap/stableswap_pool.proto:78]:
```suggestion
  repeated osmosis.gamm.stableswap.v1beta1.PoolAsset pool_assets = 6
```

**Comment by p0mvn** [comment]
@ValarDragon any updates on this?

**Comment by mattverse** [comment]
Stale bot is indeed doing its work! Nice :)

**Pull request closed without merging**